frame/support/src/traits.rs

```diff
@@ -1518,6 +1518,23 @@ pub trait OnFinalize<BlockNumber> {
 	fn on_finalize(_n: BlockNumber) {}
 }
 
+pub trait OnIdle<BlockNumber> {
+	/// The block is being finalized. Implement to have something happen.
+	///
+	/// NOTE: This function is called AFTER ALL extrinsics in a block are applied,
+	/// including inherent extrinsics.
+	fn on_idle(_n: BlockNumber, _remaining_weight: crate::weights::Weight) -> crate::weights::Weight { 0 }
+}
+
+#[impl_for_tuples(30)]
+impl<BlockNumber: Clone> OnIdle<BlockNumber> for Tuple {
+	fn on_idle(_n: BlockNumber,  _remaining_weight: crate::weights::Weight) -> crate::weights::Weight {
+		let mut weight = 0;
+		for_tuples!( #( weight = weight.saturating_add(Tuple::on_idle(_n.clone(),  _remaining_weight.clone())); )* );
+		weight
+	}
+}
```

Comment:
The tuple implementations should not  pass in the same weight. The implementation should subtract from the remaining weight after each call. Otherwise you have `n` weight to spare but `no_pallets * n` will be used.

Reply:
good catch makes sense, one thing I am thinking about now is that I can do this 

```
_remaining_weight.clone().saturating_sub(weight)
```

but can't a hook exceed the max weight for a block? So could this overflow, and isn't it ok for a hook to exceed the max weight, so checked_sub would also be an issue?

Reply:
`saturating_sub` does not modify `self`. It returns the new value. So you need an assignment to `remaining_weight`. Also you should remove the `_` prefix from your arguments. That is a convention that indicates that the arguments aren't used by the function.